src/rebar_app_discover.erl

```diff
@@ -51,24 +77,46 @@ merge_deps(AppInfo, State) ->
                  rebar_state:apply_profiles(
                    rebar_state:new(reset_hooks(rebar_state:opts(State, Default)), C,
                                   rebar_app_info:dir(AppInfo)), CurrentProfiles), Name),
+    AppState1 = rebar_state:overrides(AppState, rebar_state:get(AppState, overrides, [])),
 
-    rebar_utils:check_min_otp_version(rebar_state:get(AppState, minimum_otp_vsn, undefined)),
-    rebar_utils:check_blacklisted_otp_versions(rebar_state:get(AppState, blacklisted_otp_vsns, [])),
+    rebar_utils:check_min_otp_version(rebar_state:get(AppState1, minimum_otp_vsn, undefined)),
+    rebar_utils:check_blacklisted_otp_versions(rebar_state:get(AppState1, blacklisted_otp_vsns, [])),
 
-    AppState1 = rebar_state:set(AppState, artifacts, []),
-    AppInfo1 = rebar_app_info:state(AppInfo, AppState1),
+    AppState2 = rebar_state:set(AppState1, artifacts, []),
+    AppInfo1 = rebar_app_info:state(AppInfo, AppState2),
 
     State1 = lists:foldl(fun(Profile, StateAcc) ->
-                                 AppProfDeps = rebar_state:get(AppState, {deps, Profile}, []),
-                                 TopLevelProfDeps = rebar_state:get(StateAcc, {deps, Profile}, []),
-                                 ProfDeps2 = rebar_utils:tup_dedup(rebar_utils:tup_umerge(
-                                                     rebar_utils:tup_sort(TopLevelProfDeps)
-                                                     ,rebar_utils:tup_sort(AppProfDeps))),
-                                 rebar_state:set(StateAcc, {deps, Profile}, ProfDeps2)
+                                 handle_profile(Profile, Name, AppState1, StateAcc)
                          end, State, lists:reverse(CurrentProfiles)),
 
     {AppInfo1, State1}.
 
+handle_profile(Profile, Name, AppState, State) ->
+    {TopSrc, TopPkg} = rebar_state:get(State, {parsed_deps, Profile}, {[], []}),
+    TopLevelProfileDeps = rebar_state:get(State, {deps, Profile}, []),
+    AppProfileDeps = rebar_state:get(AppState, {deps, Profile}, []),
+    AppProfileDeps2 = rebar_utils:tup_dedup(rebar_utils:tup_sort(AppProfileDeps)),
+    ProfileDeps2 = rebar_utils:tup_dedup(rebar_utils:tup_umerge(
+                                           rebar_utils:tup_sort(TopLevelProfileDeps)
+                                           ,rebar_utils:tup_sort(AppProfileDeps2))),
+    State1 = rebar_state:set(State, {deps, Profile}, ProfileDeps2),
+
+    %% Only deps not also specified in the top level config need
+    %% to be included in the parsed deps
+    NewDeps = ProfileDeps2 -- TopLevelProfileDeps,
+    {ParsedSrc, ParsedPkg} = parse_profile_deps(Profile, Name, NewDeps, AppState, State1),
+    rebar_state:set(State1, {parsed_deps, Profile}, {TopSrc++ParsedSrc, TopPkg++ParsedPkg}).
+
+parse_profile_deps(Profile, Name, Deps, AppState, State) ->
+    DepsDir = rebar_prv_install_deps:profile_dep_dir(State, Profile),
+    Locks = rebar_state:get(State, {locks, Profile}, []),
+    rebar_prv_install_deps:parse_deps(Name
```

Comment:
We should eventually move these functions to a different module altogether. rebar_deps_utils or something